### PR TITLE
MAINT: Legacy mode specified as string, fix all-zeros legacy bug

### LINKED ISSUE #10020: np.set_printoptions(legacy=True) is not future proof
It's possible we'll want to change the repr again in future, but this is hard to extend in a backwards compatiable way. Should we use `legacy=(1,13)` or `legacy='1.13'` instead?

### PR DESCRIPTION
**Pull request by ahaldane**
Fixes #10020 and #10026

The first commit makes the `legacy` option specify version number, as `np.set_printoptions(legacy='1.13')`. 

The second commit fixes a bug where the legacy mode spacing for float arrays didn't get activated for arrays which didn't have any non-zero finite values.

### CODE REVIEW

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -87,6 +87,9 @@ def _make_options_dict(precision=None, threshold=None, edgeitems=None,
     if sign not in [None, '-', '+', ' ']:
         raise ValueError("sign option must be one of ' ', '+', or '-'")
 
+    if legacy not in [None, '', '1.13']:
+        raise ValueError("legacy option can currently only be '1.13' or ''")
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: I think this should warn, not error - otherwise, code which does not want 1.15 behavior (passing `legacy='1.14'`), will crash when run on 1.13.

**Comment by eric-wieser** [comment]
Unsure if we should be using strings or tuples here. Tuples would work better for version comparison, but perhaps that's not something we want to encourage.

**Comment by lesteve** [comment]
> Unsure if we should be using strings or tuples here. Tuples would work better for version comparison, but perhaps that's not something we want to encourage.

My 2c: use strings and `distutils.version.LooseVersion`. As long as you are comparing released versions, i.e. X.Y.Z (and not things like `1.14.dev0` or `1.14.1b1`) ` distutils.LooseVersion` is good enough. You can also  `LooseVersion('1.1.0').version` to get a list of integers corresponding to the major/minor/micro versions.

**Comment by charris** [comment]
@lesteve We don't need to worry about actual versions, just need a way to specify which printing style we want. I'd go for a string.

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (charris) [reply numpy/core/arrayprint.py]: Agree.

**Comment by ahaldane** [comment]
You're right to warn, nor raise.

Fixed.

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -87,6 +87,10 @@ def _make_options_dict(precision=None, threshold=None, edgeitems=None,
     if sign not in [None, '-', '+', ' ']:
         raise ValueError("sign option must be one of ' ', '+', or '-'")
 
+    if legacy not in [None, '', '1.13']:
+        warnings.warn("legacy printing option can currently only be '1.13' or "
+                      "''", stacklevel=3)
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Empty string seems weird to me. Can we use `False` instead?

File: numpy/core/arrayprint.py [hunk 3]
```diff
@@ -704,8 +710,8 @@ def fillFormat(self, data):
         with errstate(all='ignore'):
             hasinf = isinf(data)
             special = isnan(data) | hasinf
-            valid = not_equal(data, 0) & ~special
-            non_zero = data[valid]
+            non_special = data[~special]
+            non_zero = non_special[non_special != 0]
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Some comments might be nice here explaining why each of these are needed.

**Comment by ahaldane** [comment]
Thanks for the suggestions. I made `legacy=False` disable legacy mode, and I did a lot of cleanup in the FloatingFormat code in a new commit.

File: numpy/core/arrayprint.py [hunk 4]
```diff
@@ -841,7 +834,7 @@ def __init__(self, *args, **kwargs):
 def format_float_scientific(x, precision=None, unique=True, trim='k',
                             sign=False, pad_left=None, exp_digits=None):
     """
-    Format a floating-point scalar as a string in scientific notation.
+    Format a floating-point scalar as a decimal string in scientific notation.
```
Comment (ahaldane) [inline numpy/core/arrayprint.py]: this is an unrelated small doc tweak I am piggybacking on this PR

File: numpy/core/arrayprint.py (hunk 3 above)
Comment (charris) [reply numpy/core/arrayprint.py]: A 'not' prefix has more force than 'non'. Not worth worrying about here, just a comment.

Comment (eric-wieser) [reply numpy/core/arrayprint.py]: I would reserve a `not_` prefix for a boolean type, so `non_zero = data[not_zero]`

File: numpy/core/arrayprint.py [hunk 5]
```diff
@@ -525,11 +530,11 @@ def array2string(a, max_line_width=None, precision=None,
...
-    legacy : boolean, optional
-        If True, enables legacy printing mode, which overrides the `sign`
-        option.  Legacy printing mode approximates numpy 1.13 print output,
-        which includes a space in the sign position of floats and different
-        behavior for 0d arrays.
+    legacy : string or `False`, optional
+        If set to the string `'1.13'` enables 1.13 legacy printing mode. This
+        approximates numpy 1.13 print output by including a space in the sign
+        position of floats and different behavior for 0d arrays. If set to
+        `False`, disables legacy mode.
```
Comment (charris) [inline numpy/core/arrayprint.py]: Needs `.. versionadded::`. If folks want to use this they will need to check the NumPy version.

File: numpy/core/arrayprint.py [hunk 6]
```diff
@@ -701,40 +707,32 @@ def __init__(self, data, precision, floatmode, suppress_small, sign=False, **kwa
...
-                if not self.suppress_small and (min_val < 0.0001
-                                           or max_val/min_val > 1000.):
-                    self.exp_format = True
-
-        if len(non_zero) == 0:
+        # only the finite values are used to compute the number of digits
+        finite_vals = data[isfinite(data)]
+
+        # choose exponential mode based on the non-zero finite values:
+        abs_non_zero = absolute(finite_vals[finite_vals != 0])
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Way clearer now :)

Comment (charris) [reply numpy/core/arrayprint.py]: @eric-wieser If you are good with this, I'll make the fix and merge.

File: numpy/core/arrayprint.py (hunk 5 above)
Comment (eric-wieser) [reply numpy/core/arrayprint.py]: Might want a comment here explaining that invalid versions will be ignored for forward-compatibility too.

Comment (eric-wieser) [reply numpy/core/arrayprint.py]: Don't forget to add this to both docstrings

File: numpy/core/arrayprint.py [hunk 7]
```diff
@@ -701,40 +707,32 @@ def __init__(self, data, precision, floatmode, suppress_small, sign=False, **kwa
         self.fillFormat(data)
 
     def fillFormat(self, data):
-        with errstate(all='ignore'):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Why did this line go away?

**Comment by eric-wieser** [comment]
A little puzzled by the `errstate(all='ignore')` vanishing, but all looks good to me modulo the doc fix you spotted.

File: numpy/core/arrayprint.py (hunk 7 above)
Comment (ahaldane) [reply numpy/core/arrayprint.py]: Oh I think I need to keep it. It must be because the division operation can overflow, eg `1.0/np.float64(2.2250738585072e-310)`.

Comment (charris) [reply numpy/core/arrayprint.py]: Oops, I just made the other fixes online.

Comment (charris) [reply numpy/core/arrayprint.py]: You may need to update from origin.

Comment (eric-wieser) [reply numpy/core/arrayprint.py]: Maybe push it to a deeper scope this time

File: numpy/core/tests/test_arrayprint.py [hunk 8]
```diff
@@ -354,6 +354,11 @@ def test_sign_spacing(self):
 
         assert_raises(TypeError, np.set_printoptions, wrongarg=True)
 
+    def test_float_overflow_nowarn(self):
+        # make sure internal computations in FloatingFormat don't
+        # warn about overflow
+        repr(array([1.0, 1e-310], dtype='f8'))
```
Comment (eric-wieser) [inline numpy/core/tests/test_arrayprint.py]: Nice!

**Review by eric-wieser — APPROVED** [review]
LGTM. Might be nice to squash @charris' commits together

File: numpy/core/tests/test_arrayprint.py (hunk 8 above)
Comment (ahaldane) [reply numpy/core/tests/test_arrayprint.py]: Hmm but actually tests just finished running on my system and I don't think this triggers the problem because the `min_val < 0.0001` short-circtuits out the division clause anytime overflow could occur.

I think overflow may only be posible for float16 types under certain circumstances. `np.array([1e4, 0.1], dtype='f2')` works.

**Comment by charris** [comment]
The `errstate` block seems to have originated back around 2007, although not implemented that way as `errstate` was a later addition.

**Comment by charris** [comment]
Thanks Allan.
